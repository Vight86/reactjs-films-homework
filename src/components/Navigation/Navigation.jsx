import React, { useState } from 'react';
import propTypes from 'prop-types';
import { NavLink, useRouteMatch } from 'react-router-dom';
import SelectButton from '../SelectButton/index';
import styles from './Navigation.scss';

const Navigation = ({ genres }) => {
  const match = useRouteMatch('/:category/movie/:id');
  let id = match?.params?.id;
  id = id ? `/movie/${id}` : '';

  const [isMobMenuOpen, setIsMobMenuOpen] = useState(false);
  const handleClick = () => {
    setIsMobMenuOpen(!isMobMenuOpen);
  };

  return (
    <>
      <button
        type="button"
        className={`${styles.mobileMenuButton} ${isMobMenuOpen ? styles.active : ''}`.trim()}
        onClick={handleClick}
      >
        <svg viewBox="0 0 32 7" className={styles.line} preserveAspectRatio="none">
          <line x1={4} y1={4} x2={28} y2={4} />
        </svg>
        <svg viewBox="0 0 32 7" className={styles.line} preserveAspectRatio="none">
          <line x1={4} y1={4} x2={28} y2={4} />
        </svg>
        <svg viewBox="0 0 32 7" className={styles.line} preserveAspectRatio="none">
          <line x1={4} y1={4} x2={28} y2={4} />
        </svg>
      </button>
      <nav className={`${styles.navigation} ${isMobMenuOpen ? styles.active : ''}`.trim()}>
        <NavLink
          to={`/popular${id}`}
          className={styles.navItem}
          activeClassName={styles.navItem_active}
          data-text="Trending"
          onClick={handleClick}
        >
          Trending
        </NavLink>
        <NavLink
          to={`/top_rated${id}`}
          className={styles.navItem}
          activeClassName={styles.navItem_active}
          data-text="Top rated"
          onClick={handleClick}
        >
          Top rated
        </NavLink>
        <NavLink
          to={`/upcoming${id}`}
          className={styles.navItem}
          activeClassName={styles.navItem_active}
          data-text="ComingSoon"
          onClick={handleClick}
        >
          Coming soon
        </NavLink>
        <SelectButton
          genres={genres}
          onClick={handleClick}
        >
          Genre
        </SelectButton>
      </nav>
    </>
  );
};

Navigation.propTypes = {
  genres: propTypes.arrayOf(propTypes.object).isRequired,
};

export default Navigation;
