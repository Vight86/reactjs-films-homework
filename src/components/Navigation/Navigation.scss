@import '../../utils/variables';

.navigation {
  display: flex;
  align-items: center;
  padding: 0;
  margin: 0 0 -1px 0;
  list-style: none;

  @media screen and (max-width: $bp-xxxsmall) {
    position: absolute;
    top: 48px;
    width: 100%;
    box-sizing: border-box;
    flex-direction: column;
    align-items: flex-start;
    padding: 10px;
    background: $bg-color;
    opacity: 0;
    pointer-events: none;
    transform: translateX(-100%);
    transition: all 0.3s ease-out;
    visibility: hidden;

    &.active {
      opacity: 1;
      pointer-events: all;
      transform: translateX(0);
      visibility: visible;
    }
  }
}

.navItem {
  display: flex;
  flex-direction: column;
  padding-bottom: 0.4em;
  border: none;
  border-bottom: 4px solid transparent;
  margin-right: 1.8em;
  background: none;
  color: $color-secondary;
  cursor: pointer;
  font-family: $font-primary;
  font-size: 2.8em;
  letter-spacing: 0.04em;
  text-decoration: none;
  transition: all 0.2s;

  @media screen and (max-width: $bp-xxxsmall) {
    width: 100%;
    margin-bottom: 10px;
  }

  &::after {
    overflow: hidden;
    height: 0;
    content: attr(data-text);
    content: attr(data-text) / "";
    font-weight: 700;
    pointer-events: none;
    user-select: none;
    visibility: hidden;
  }

  &_active {
    border-bottom: 4px solid currentColor;
    font-weight: 700;
    outline: none;
  }
}

.line {
  position: absolute;
  left: 0;
  stroke: $color-secondary;
  stroke-linecap: round;
  stroke-width: 5;
  transition: transform 0.3s cubic-bezier(0.64, 0.76, 0.69, 1.76);

  &:nth-child(1) {
    top: 0;
  }

  &:nth-child(2) {
    top: calc(50% - 2px);
  }

  &:nth-child(3) {
    top: calc(100% - 4px);
  }
}

.mobileMenuButton {
  position: relative;
  display: none;
  width: 40px;
  height: 30px;
  padding: 0;
  border: none;
  background: none;
  cursor: pointer;

  @media screen and (max-width: $bp-xxxsmall) {
    display: block;
  }

  &.active {
    .line:nth-child(1) {
      transform: translateX(25%);
    }

    .line:nth-child(2) {
      transform: translateX(40%);
    }

    .line:nth-child(3) {
      transform: translateX(60%);
    }
  }
}
