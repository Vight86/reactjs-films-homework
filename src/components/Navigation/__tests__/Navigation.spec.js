import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Navigation from '../index';

let mockUseRouteMatchObject;
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: jest.fn(() => mockUseRouteMatchObject),
}));

let mockGenres;
let renderer;
let result;

beforeEach(() => {
  mockUseRouteMatchObject = { params: { id: '111' } };
  mockGenres = [{ id: 1, name: 'Action' }];
  renderer = new ShallowRenderer();
  renderer.render(<Navigation genres={mockGenres} />);
  result = renderer.getRenderOutput();
});

afterEach(() => {
  mockUseRouteMatchObject = null;
  mockGenres = null;
  renderer = null;
  result = null;
});

describe('render Navigation component', () => {
  it('render component correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('render if MovieDetailsPage is opened', () => {
    mockUseRouteMatchObject = null;
    renderer.render(<Navigation genres={mockGenres} />);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render mobile menu button clicks', () => {
    const mobMenuButton = result.props.children[0];
    const { onClick } = mobMenuButton.props;
    onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
