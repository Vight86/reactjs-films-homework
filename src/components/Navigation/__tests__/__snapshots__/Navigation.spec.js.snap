// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Navigation component render component correctly 1`] = `
<React.Fragment>
  <button
    className="mobileMenuButton"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
  </button>
  <nav
    className="navigation"
  >
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Trending"
      onClick={[Function]}
      to="/popular/movie/111"
    >
      Trending
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Top rated"
      onClick={[Function]}
      to="/top_rated/movie/111"
    >
      Top rated
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="ComingSoon"
      onClick={[Function]}
      to="/upcoming/movie/111"
    >
      Coming soon
    </NavLink>
    <SelectButton
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
        ]
      }
      onClick={[Function]}
    >
      Genre
    </SelectButton>
  </nav>
</React.Fragment>
`;

exports[`render Navigation component render if MovieDetailsPage is opened 1`] = `
<React.Fragment>
  <button
    className="mobileMenuButton"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
  </button>
  <nav
    className="navigation"
  >
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Trending"
      onClick={[Function]}
      to="/popular"
    >
      Trending
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Top rated"
      onClick={[Function]}
      to="/top_rated"
    >
      Top rated
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="ComingSoon"
      onClick={[Function]}
      to="/upcoming"
    >
      Coming soon
    </NavLink>
    <SelectButton
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
        ]
      }
      onClick={[Function]}
    >
      Genre
    </SelectButton>
  </nav>
</React.Fragment>
`;

exports[`render Navigation component render mobile menu button clicks 1`] = `
<React.Fragment>
  <button
    className="mobileMenuButton active"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
    <svg
      className="line"
      preserveAspectRatio="none"
      viewBox="0 0 32 7"
    >
      <line
        x1={4}
        x2={28}
        y1={4}
        y2={4}
      />
    </svg>
  </button>
  <nav
    className="navigation active"
  >
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Trending"
      onClick={[Function]}
      to="/popular/movie/111"
    >
      Trending
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="Top rated"
      onClick={[Function]}
      to="/top_rated/movie/111"
    >
      Top rated
    </NavLink>
    <NavLink
      activeClassName="navItem_active"
      className="navItem"
      data-text="ComingSoon"
      onClick={[Function]}
      to="/upcoming/movie/111"
    >
      Coming soon
    </NavLink>
    <SelectButton
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
        ]
      }
      onClick={[Function]}
    >
      Genre
    </SelectButton>
  </nav>
</React.Fragment>
`;
